File: app/src/main/java/com/semisonfire/cloudgallery/ui/disk/adapter/PhotoAdapter.java
package com.semisonfire.cloudgallery.ui.disk.adapter;

import android.support.annotation.NonNull;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.semisonfire.cloudgallery.R;
import com.semisonfire.cloudgallery.data.model.Photo;
import com.semisonfire.cloudgallery.ui.custom.PhotoDiffUtil;
import com.semisonfire.cloudgallery.ui.custom.SelectableHelper;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.List;

public class PhotoAdapter extends RecyclerView.Adapter<PhotoAdapter.PhotoViewHolder> {

    private SelectableHelper.OnPhotoListener photoListener;
    private List<Photo> photos = new ArrayList<>();

    public PhotoAdapter() {
    }

    public void setPhotoListener(SelectableHelper.OnPhotoListener photoListener) {
        this.photoListener = photoListener;
    }

    @NonNull
    @Override
    public PhotoViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_photo, parent, false);
        return new PhotoViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull PhotoViewHolder holder, int position) {
        holder.bind(photos.get(position));
    }

    @Override
    public int getItemCount() {
        return photos.size();
    }

    public void setPhotos(List<Photo> items) {
        final PhotoDiffUtil diffUtilCallback = new PhotoDiffUtil(items, photos);
        final DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(diffUtilCallback);
        photos.clear();
        photos.addAll(items);
        diffResult.dispatchUpdatesTo(this);
    }

    public void addPhotos(List<? extends Photo> photos) {
        this.photos.addAll(photos);
        notifyItemRangeInserted(getItemCount(), photos.size());
    }

    public void setSelection(boolean selected) {
        if (!selected) {
            for (Photo photo : photos) {
                if (photo.isSelected()) {
                    photo.setSelected(false);
                }
            }
        }
        SelectableHelper.setMultipleSelection(selected);
        notifyDataSetChanged();
    }

    public void remove(Photo photo) {
        int position = photos.indexOf(photo);
        photos.remove(photo);
        notifyItemRemoved(position);
    }

    class PhotoViewHolder extends RecyclerView.ViewHolder {

        private int targetHeight;
        private int targetWidth;
        private ImageView photoImage;
        private ImageView selectImage;

        public PhotoViewHolder(View itemView) {
            super(itemView);
            photoImage = itemView.findViewById(R.id.image_photo);
            selectImage = itemView.findViewById(R.id.image_selected);
            targetWidth = itemView.getContext().getResources().getDimensionPixelOffset(R.dimen.photo_max_width);
            targetHeight = itemView.getContext().getResources().getDimensionPixelOffset(R.dimen.photo_max_height);
        }

        void bind(Photo photo) {

            View.OnClickListener onItemClick = v -> {
                if (photoListener != null) {
                    Photo p = photos.get(getAdapterPosition());
                    if (!SelectableHelper.getMultipleSelection()) {
                        photoListener.onPhotoClick(photos, getAdapterPosition());
                    } else {
                        p.setSelected(!p.isSelected());
                        notifyItemChanged(getAdapterPosition());
                        photoListener.onSelectedPhotoClick(p);
                    }
                }
            };

            photoImage.setImageDrawable(null);
            Picasso.get().load(photo.getPreview())
                    .resize(targetWidth, targetHeight)
                    .centerCrop()
                    .placeholder(R.color.black)
                    .error(R.drawable.ic_gallery)
                    .into(photoImage);

            photoImage.setOnClickListener(onItemClick);
            selectImage.setOnClickListener(onItemClick);
            selectImage.setVisibility(photo.isSelected() ? View.VISIBLE : View.GONE);

            photoImage.setOnLongClickListener(v -> {
                if (photoListener != null && !SelectableHelper.getMultipleSelection()) {
                    Photo p = photos.get(getAdapterPosition());
                    p.setSelected(true);
                    notifyItemChanged(getAdapterPosition());
                    photoListener.onPhotoLongClick();
                    photoListener.onSelectedPhotoClick(p);
                    return true;
                }
                return false;
            });
        }
    }
}

File: app/src/main/java/com/semisonfire/cloudgallery/ui/disk/adapter/PhotoAdapter.kt
package com.semisonfire.cloudgallery.ui.disk.adapter

import android.view.View
import android.widget.ImageView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.semisonfire.cloudgallery.R
import com.semisonfire.cloudgallery.core.data.model.Photo
import com.semisonfire.cloudgallery.core.ui.adapter.BaseAdapter
import com.semisonfire.cloudgallery.core.ui.adapter.BaseViewHolder
import com.semisonfire.cloudgallery.image.utils.loadRoundedImage
import com.semisonfire.cloudgallery.ui.custom.PhotoDiffUtil
import com.semisonfire.cloudgallery.ui.custom.SelectableHelper
import com.semisonfire.cloudgallery.ui.custom.SelectableHelper.OnPhotoListener
import com.semisonfire.cloudgallery.ui.disk.adapter.PhotoAdapter.PhotoViewHolder
import com.semisonfire.cloudgallery.utils.dimen

class PhotoAdapter : BaseAdapter<Photo, PhotoViewHolder>() {

    private var photoListener: OnPhotoListener? = null

    fun setPhotoListener(photoListener: OnPhotoListener) {
        this.photoListener = photoListener
    }

    override fun layoutId(): Int {
        return R.layout.item_photo
    }

    override fun createViewHolder(view: View): PhotoViewHolder {
        val viewHolder = PhotoViewHolder(view)
        addClickListener(viewHolder)
        return viewHolder
    }

    private fun addClickListener(photoViewHolder: PhotoViewHolder) {
        val itemClick = View.OnClickListener {
            val adapterPosition = photoViewHolder.adapterPosition
            if (adapterPosition == RecyclerView.NO_POSITION) return@OnClickListener

            val photo = items[adapterPosition]
            if (!SelectableHelper.getMultipleSelection()) {
                photoListener?.onPhotoClick(items, adapterPosition)
            } else {
                photo.isSelected = !photo.isSelected
                notifyItemChanged(adapterPosition)
                photoListener?.onSelectedPhotoClick(photo)
            }
        }
        photoViewHolder.photoImage.setOnClickListener(itemClick)
        photoViewHolder.photoImage.setOnLongClickListener {
            val adapterPosition = photoViewHolder.adapterPosition
            if (adapterPosition == RecyclerView.NO_POSITION) return@setOnLongClickListener false

            if (photoListener != null && !SelectableHelper.getMultipleSelection()) {
                val photo = items[adapterPosition]
                photo.isSelected = true
                notifyItemChanged(adapterPosition)

                photoListener?.onPhotoLongClick()
                photoListener?.onSelectedPhotoClick(photo)
                return@setOnLongClickListener true
            }
            false
        }
    }

    override fun updateDataSet(newItems: List<Photo>) {
        val diffUtilCallback = PhotoDiffUtil(newItems, items)
        val diffResult = DiffUtil.calculateDiff(diffUtilCallback)
        items.clear()
        items.addAll(newItems)
        diffResult.dispatchUpdatesTo(this)
    }

    fun setSelection(selected: Boolean) {
        if (!selected) {
            for (photo in items) {
                if (photo.isSelected) {
                    photo.isSelected = false
                }
            }
        }
        SelectableHelper.setMultipleSelection(selected)
        notifyDataSetChanged()
    }

    class PhotoViewHolder(itemView: View) : BaseViewHolder<Photo>(itemView) {

        val photoImage: ImageView = itemView.findViewById(R.id.image_photo)
        private val selectImage: ImageView = itemView.findViewById(R.id.image_selected)

        private val targetHeight = itemView.context.dimen(R.dimen.photo_max_height)
        private val targetWidth = itemView.context.dimen(R.dimen.photo_max_width)

        override fun bindItem(item: Photo) {
            photoImage.setImageDrawable(null)
            photoImage.loadRoundedImage(item.preview, targetWidth, targetHeight)
            selectImage.visibility = if (item.isSelected) View.VISIBLE else View.GONE
        }
    }
}